### PR TITLE
Add option to reconcile the account(s) after an import, if a balance …

### PR DESCRIPTION
**Pull request by jeanlaroche**
…is available.

The user has an option to allow/disallow this and the setting is saved in the
preferences.

### CODE REVIEW

**Comment by Bob-IT** [comment]
@jeanlaroche A couple of observations...
Your master and maint branches are out of date so would suggest that you keep them updated.
The fix for the 'seg-fault' in the import-main-matcher now appears in two PR's so I would suggest that you create a new PR with just the changes to that file. Please take care in grabbing the changes in PR697 and this one as the unregister position is correct in this PR but the other PR has the other parts correct. Once that is done it can be dropped from this one and the other.

As you will see, Travis has failed, do you run this on you github?, I have mentioned the offending line inline along with some others.

File: gnucash/gschemas/org.gnucash.warnings.gschema.xml.in [hunk 1]
```diff
@@ -99,6 +99,11 @@
       <summary>Commit changes to a transaction</summary>
       <description>This dialog is presented when you attempt to move out of a modified transaction. The changed data must be either saved or discarded.</description>
     </key>
+    <key name="rec-after-match" type="i">
+      <default>0</default>
+      <summary>Reconcile after matching transactions</summary>
+      <description>This dialog is presented when you are done matching transactions and there is a balance information in your statement.</description>
```
Comment (Bob-IT) [inline gnucash/gschemas/org.gnucash.warnings.gschema.xml.in:105]: I think sentence needs changing, drop the 'a' so 'and there is balance'
Also below.

File: gnucash/import-export/ofx/gnc-ofx-import.c [hunk 2]
```diff
@@ -1038,6 +1039,54 @@ double ofx_get_investment_amount(const struct OfxTransactionData* data)
...
+            {
+                // Now's the time to ask the user (but only once)
+                if (response == GTK_RESPONSE_NONE)
+                {
+                    response =  gnc_dialog_run (GTK_DIALOG(dialog), warning);
+                    if (response != GTK_RESPONSE_ACCEPT)
+                        break;
+                }
+                gnc_numeric value = double_to_gnc_numeric (statement->ledger_balance, xaccAccountGetCommoditySCU (account), GNC_HOW_RND_ROUND_HALF_UP);
+                recnWindowWithBalance (GTK_WIDGET (info->parent), account, value, statement->ledger_balance_date);
```
Comment (Bob-IT) [inline gnucash/import-export/ofx/gnc-ofx-import.c:1081]: This is the travis failure, add 'gnc_numeric value;' to the start of the if block and change this line to 'value = ...'

File: gnucash/import-export/ofx/gnc-ofx-import.c [hunk 3]
```diff
@@ -1038,6 +1039,54 @@ double ofx_get_investment_amount(const struct OfxTransactionData* data)
...
+        return;
+    
+    if (response_id == GTK_RESPONSE_OK)
+    {
+        gint response = GTK_RESPONSE_NONE;
+        // Prepare a dialog to ask the user whether they want to open a reconcile window.
+        GtkWidget* dialog = gtk_message_dialog_new ( info->parent,
+                                                    GTK_DIALOG_MODAL | GTK_DIALOG_DESTROY_WITH_PARENT,
+                                                    GTK_MESSAGE_WARNING, GTK_BUTTONS_NONE,
+                                                    "%s", _("Your statement(s) include(s) balance information,\ndo you want to reconcile now?"));
```
Comment (Bob-IT) [inline gnucash/import-export/ofx/gnc-ofx-import.c:1059]: Personally I do not like this, can it be changed to ngettext(3) message so you have "Your statement includes..." and "Your statements include..."

**Comment by Bob-IT** [comment]
Bearing in mind I do not use this...
Also the I do not think the warning box is functioning the way you expect, if I try an import and then press 'OK' the message box pops up with yes/no. So as I do not want to ever go to reconcile, I tick 'Remember the answer and don't ask me again' and press 'NO'. That does not keep the setting, it only when I select 'YES' it remembers so you may have to change the semantics to get what you want.
Did you consider adding another button that did the OK procedure and then started the reconcile dialog, not sure what the wording would be but maybe something like 'Complete matcher, Reconcile Now' and you could probably control the sensitivity based on balance present.

**Comment by jeanlaroche** [comment]
Apologies for all this mess. These PRs are intermingled, I fixed the issues you and others mentioned, then in the process of merging with other changes, I undid them. :( Sorry about that.
And no, I don't run travis on my repo, and I should, I didn't realize it was possible.

**Comment by jeanlaroche** [comment]
> Did you consider adding another button that did the OK procedure and then started the reconcile dialog, not sure what the wording would be but maybe something like 'Complete matcher, Reconcile Now' and you could probably control the sensitivity based on balance present.

I'm not sure I understand what your mean here... Currently, you do the matching, then when that's done, you get a choice to reconcile or not. I don't quite get what that other button would do...

**Comment by Bob-IT** [comment]
> I'm not sure I understand what your mean here... Currently, you do the matching, then when that's done, you get a choice to reconcile or not. I don't quite get what that other button would do..

This was just a thought instead of you callback.

**Comment by Bob-IT** [comment]
As stated above, I do not use this.
I think the code looks OK and does what is intended.
@christopherlam Do you have any thoughts on this ?
It does what is required but my only reservation is when the user says 'No' and selects remember this, are they ever going to turn it back on?

File: gnucash/import-export/ofx/gnc-ofx-import.c [hunk 4]
```diff
@@ -1038,6 +1039,55 @@ double ofx_get_investment_amount(const struct OfxTransactionData* data)
...
+    ofx_info* info = (ofx_info*) user_data;
+    GSList *iter = NULL;
+    
+    if (info->statement == NULL)
+        return;
+    
+    if (response_id == GTK_RESPONSE_OK)
+    {
+        char* message = ngettext ("Your statement includes balance information\nDo you want to reconcile now?",
+                                  "Your statements include balance information\nDo you want to reconcile now?",
```
Comment (christopherlam) [inline gnucash/import-export/ofx/gnc-ofx-import.c:1055]: Punctuation: `Your statement includes ending balances. \n\nDo you want to reconcile now?`.

@gjanssens likes to minimise use of `gnc-warning` preferences; personally I think this is a valid use for it.

File: gnucash/import-export/ofx/gnc-ofx-import.c [hunk 5]
```diff
@@ -1038,6 +1039,55 @@ double ofx_get_investment_amount(const struct OfxTransactionData* data)
     }
 }
 
+// This callback is called when the user is done matching transactions.
+static void
+gnc_ofx_match_done (GtkDialog *dialog, gint response_id, gpointer user_data)
+{
+    ofx_info* info = (ofx_info*) user_data;
+    GSList *iter = NULL;
+
```
Comment (christopherlam) [inline gnucash/import-export/ofx/gnc-ofx-import.c:1048]: There's trailing whitespace here and 3 lines down...

**Comment by christopherlam** [comment]
P.S. for @gjanssens to offer a review without needing to build, here are the successive screenshots: (note the _**new**_ import matcher showing matched split data \o/) (note also the desc/memo fields are empty due to malformed OFX, bug 797633)

image
image
image

**Comment by goodvibes2** [comment]
Hi @jeanlaroche, My ofx files never include a balance, so IIUC, I'll never see your question?
Also, for 1 account, I probably would want to reconcile, and another I wouldn't.
Would it be possible to add an 'OK + Reconcile' button next to the 'OK' button instead of using the new dialog. That would fix my concerns.

**Comment by jeanlaroche** [comment]
Oh I see, add a third button when you pick the file so indicate you'd like to reconcile. That's a good solution. I believe if was suggested before but I failed to click!

**Comment by goodvibes2** [comment]
@jeanlaroche, I was thinking the new button should be on the last screen where you OK the import to actually occur after the matching, as after that is when you would Reconcile if needed.
A checkbox ('Reconcile after import', default Unchecked) may be better ui but would need an extra click sometimes.

**Comment by jeanlaroche** [comment]
Oh, a new button on the match dialog, that makes more sense yes... I looked into have a check box for "reconcile when done" but that was not a very clean implementation (because the matching dialog only return a standard message (Yes, No, OK, Cancel) so I'd have to create a custom return...

**Comment by christopherlam** [comment]
I was wondering whether adding another button to importer would lead to over engineering; but Jean seems to have good skills...

**Comment by Bob-IT** [comment]
> Oh, a new button on the match dialog, that makes more sense yes... I looked into have a check box for "reconcile when done" but that was not a very clean implementation (because the matching dialog only return a standard message (Yes, No, OK, Cancel) so I'd have to create a custom return...

Obviously I did not make my self clear enough...

**Comment by jeanlaroche** [comment]
I have implemented the ideas mentioned above, but they depend on my other PR (importing multiple OFX files) so I'm going to wait for that one to clear before I push the new code here. I"m not sure what the proper procedure is in this type of situation...

File: gnucash/import-export/ofx/gnc-ofx-import.c (hunk 4 above)
Comment (gjanssens) [reply gnucash/import-export/ofx/gnc-ofx-import.c:1055]: Agreed :)

**Pull request closed without merging**